### ISSUE TITLE
Custom translation overrides

### ISSUE THREAD

**Issue by lache-melvin** [msupply-foundation/open-msupply#5061]
## Is your feature request related to a problem? Please describe 👀

A way for users to configure custom translations on the OMS central server. 

The user should be able to:
- Override existing translation strings
- Add new translation keys, e.g. for use in custom JSON form schemas

## Describe the solution you'd like 🎁

Coupled together with global preferences: 
- Stored in mutex in service provider for report translations
- Cached value is served to frontend

QUESTIONS:
- Do we need to support custom translations for multiple languages? Probably more complete if we do

### Describe alternatives you've considered 💭

- Key value store?
- Custom translations override all languages, as opposed to per language
- Translation editor - with more feedback about available translations, changes to common.json translations, keys that have inputs e.g. `{{ count }}`

### Moneyworks Jobcode 🧰
OMS:NGUCF

## Agreed Solution

Overrides will apply for all languages

- [x] Store the overrides as a Global Preference of serialised JSON
- [x] Preference editor is a table, in a modal (therefore recommend custom PreferenceValueType, to map to renderer)
  - [x] Shows the translation key, default translation, and your custom override
  - [x] You can search by default translation to add new overrides (maybe can search by key as well?)
  - [x] Delete lines
  - [x] Column UI UX - add at top, sort order
  - [x] Additional tests
- [x] Custom translations are applied to frontend features
  - [x] Hopefully could use the i18n `namespace` feature, as long as we expose `custom.json` from backend?
- [x] Custom translations are applied to backend (i.e. for reports)
- [ ] Future: ability to disable custom translations being applied
- [x] Invalidate custom translation keys on sync
- [x] Add error logging for backend localisations
- [x] Add tests for backend localisations
- [x] Add validation for saving translations

**Comment by roxy-dao** [comment]
TODO: Check with report JSON form translations to make sure we're not doubling up on any work etc

**Comment by mark-prins** [comment]
> Always serve custom.json along with common.json (would be empty object by default)

to clarify: `custom.json` would be empty by default, right? The proximity made me think the brackets relate to `common.json`

**Comment by mark-prins** [comment]
Alternatively - the server combines `custom.json` and `common.json` and serves up a single file. 
This would be easier to code for on the front end as no change would be required. If we have two files then we need to either combine the json files (during the i18n initialisation? ) or modify the `useTranslation` hook to always use the `custom` namespace while using `common` as the default or ..?

> Do we need to support custom translations for multiple languages?

that would be more comprehensive as a solution, so I'd prefer to support that. In practice though - countries have only used a single language and would therefore only require a single language in the custom namespace. So we're adding complexity which would be unused.

**Comment by lache-melvin** [comment]
> modify the `useTranslation` hook to always use the `custom` namespace while using `common` as the default

Yeah that was the thinking... just seeing as the `useTranslation` hook already has the "merging" functionality baked in, why not make the most of it! Not sure if that's a little safer too, keeping `common.json` separate so we know the fallbacks are definitely available?

Though yes, combining on the server is also a nice solution, and could be shared by both frontend and backend (for reports) 👀

**Comment by lache-melvin** [comment]
Update: we're thinking let's store the custom translations with the global prefs, incoming from @msupply-foundation/ruru

- Stored in mutex in service provider for report translations
- Cached value is served to frontend

**Comment by lache-melvin** [comment]
Blocked until global prefs implemented

**Comment by lache-melvin** [comment]
Adding triage label again so this can be discussed - global prefs aren't likely to be in with enough time before RC for us to get this into 2.4... move to 2.5? Is there room?

**Comment by andreievg** [comment]
Moved to 2.5, still needs global pref

**Comment by jmbrunskill** [comment]
@andreievg can you please share your thoughts about using global prefs for this? I don't see a global prefs issue or design so keen to understand what we need to do to unblock this...

**Comment by roxy-dao** [comment]
> @andreievg can you please share your thoughts about using global prefs for this? I don't see a global prefs issue or design so keen to understand what we need to do to unblock this...

https://github.com/msupply-foundation/open-msupply/issues/5138#issuecomment-2461246641

**Comment by roxy-dao** [comment]
But please create another issue for the global pref implementation itself (:

**Comment by jmbrunskill** [comment]
@andreievg are you ok if we do this without using global prefs?
Talking with @mark-prins, we think this is a higher priority than new global prefs functionality and potentially justifies its own table/potentially serving separately.

It's not too much work to create a new sync'd table...

**Comment by andreievg** [comment]
> It's not too much work to create a new sync'd table...

Should be pretty easy with the example commits in the sync doc

Ok with new table for this, but I still feel global preferences should be done sooner then later

**Comment by roxy-dao** [comment]
@andreievg  to break it down to manageable issues

**Comment by jmbrunskill** [comment]
Developers to propose the technical details, make sure to run it past @andreievg before too much work is done.

**Comment by CarlosNZ** [comment]
Another use case: #7143

**Comment by fergie-nz** [comment]
I am thinking a common use case would also be back end translation overrides for any JSON forms - or maybe a user accessible way to change a standard report without requesting a custom report.

There are embedded.

**Comment by mark-prins** [comment]
Global prefs are in which was the reason for the blocker.
Have added triage so that we can discuss and allocate to a milestone

**Comment by andreievg** [comment]
We would love to do this in 2.8, but only if current 2.8 commitments are completed. Putting as high severity for cooldown, if all 2.8 tasks are finished, this should be the first to be brought in

**Comment by roxy-dao** [comment]
@andreievg to look at today and flesh out issue. 

1. MVP text box
2. Part of global prefs

Do we pre-load their current language `.json` file into the box? How to make sure they get the correct syntax?

**Comment by roxy-dao** [comment]
I'll keep issue opened until 

- Variables in translations (count etc). Validate that the var name is accurate, and stop user from saving. Highlight the inaccurate line.
- Show _one & _other etc together
- Typing \ on search gives Invalid regular expression: /\/i: \ at end of pattern
- Multiline edit instead of single text line. Needs to support new line
- Why is the modal height expanding until it reaches max height….

These done

**Comment by alainsussol** [comment]
Tested in feature Ruru 2.10. See Results in #8834
